File: vpo-backend/node-engine/src/io_routing.rs
use serde::{Deserialize, Serialize};

use crate::node::NodeIndex;

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
#[serde(tag = "variant", content = "data")]
pub enum DeviceType {
    Midi,
    Stream,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
#[serde(tag = "variant", content = "data")]
pub enum DeviceDirection {
    Source,
    Sink,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
#[serde(rename_all = "camelCase")]
pub struct RouteRule {
    pub device_id: String,
    pub device_type: DeviceType,
    pub device_direction: DeviceDirection,
    pub device_channel: usize,
    pub node: NodeIndex,
    pub node_socket: usize,
    pub node_channel: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct DeviceInfo {
    pub name: String,
    pub device_type: DeviceType,
    pub device_direction: DeviceDirection,
    pub channels: usize,
    pub buffer_size: usize,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct IoRoutes {
    pub rules: Vec<RouteRule>,
    pub devices: Vec<DeviceInfo>,
}

File: vpo-frontend/src/lib/node-engine/io_routing.ts
import type { DiscriminatedUnion } from "$lib/util/discriminated-union";

export type DeviceType = DiscriminatedUnion<"variant", {
    Midi: {},
    Stream: {},
}>;

export type DeviceDirection = DiscriminatedUnion<"variant", {
    Source: {},
    Sink: {},
}>;

export type RouteRule = {
    deviceId: string,
    deviceType: DeviceType,
    deviceDirection: DeviceDirection,
    deviceChannel: number,
    node: string,
    nodeChannel: number,
};

export type DeviceInfo = {
    name: string,
    deviceType: DeviceType,
    deviceDirection: DeviceDirection,
    channels: number,
    bufferSize: number,
};

export type IoRoutes = {
    rules: Array<RouteRule>,
    devices: Array<DeviceInfo>,
};